Crash opening binary file

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

## Bug
Zed crashes when opening this binary file. 

## How to reproduce
1. Copy the base64 encoding and paste it at https://base64.guru/converter/decode/file.
2. Download the resulting binary file.
3. Open the file in zed.
Then Zed should immediately crash.
 
<details><summary>base64 encoding:</summary>

```
AAAIAQAAJxAJAgEBBgEEBgUHBAUHAwQBAgQIAAIFBwkBBAYACQMICAMDCAAHBQcJBgEDBwYHAgECAgQEBQgCAggECAAHBwgFAQECAwkIBwACBgIDAQIIBAEIBQkFAAMCAAYFAwYHAQgAAQQCAwYHAgcIBQkJBAIFBwAFAggGBwgAAAkJAwAIBAEFBAEJAQgGAgECBQEAAAABBgEGAgIEBAEEBQAEBwkDBwIDCQAJBAcEAgAFAgECAQMACQEACQMGBwkJBAQHAQIBBgMCCAMGAQEAAgkCBAAHCQgEAQgEAQMBBgcCCAUCAAcHBgIHAAcICQIJAAUBBAQFBgkCBggGBAICCQcGBQUCCAUCAwAECAAABgMICQYBAwACAwAIAwcEAAEGAwAGBgcFBAUJBQYFBQEJCAgDAwYIAAACCQcHAQMGBAIECQEGBAgCBgUEBgcHBwMDBwAHAQMHAgIDBAADAQYBCQQJCQEHCAMGAAIECAABAgIEBAcDAgIFAAcJBAAJAwkEAgUAAgMFCAEGAgYGBAkFAgAEBgECAAkHAAQEBgYCAQAGBwIJBAIBBQQFAwgFCAQECAkIBgIEBAIEAQYBAwAJCAgGBQMBCQUDAwABAAIJAgAGAwQCAAADAwgCCAkEAAcGBAQJAgkFCQMHBQcIAQAAAwQICQcJAQIHAAIABQcBCAICAgIEAgYBCQgFAQkBAggDCAkCBAIICAIABQgIBQMJBAMEBAUBAAECAAYJBgEFAQEBCQMEBQMGBgYEBgAFCAMDBAYHBwgJAAMIBwYHCQMEBgYFAAEBBQkEBgUHCAECCQcHAQAJAwIJAAcGCAIHAAIDCAICBgkDAAcIBgIJBAIGBggABAQEBwUIBAkBAAUEBAQAAAQFBgAEBQQBAwEBBgQDCAIIAgcABAAIBQAIAgkACAkGBAIJAwQFAAkFAwYGCAMDCAEEAAkHCQQICQEGBwMAAgYHAQAACAUCBgYGBAEFCQACAQYFAAADAwMIAQEIBQcHCAcCAAcDCAAJBwYBAgYECAAGAwYIBAAICQIEBQkBBAQJAgEHCQUIAQcHAQEBBgkFAwgEAgcECAMCBAQHAQQJAwUIBQQHAggFCQMDBgcBBwMFBAQFCAMHAQIGAQkIAgcBAwcFCQkBCAIFBwEJCAEAAAYBBwEBBQcBBQICAwMBAQQJBAMHBwAICQkGAQMEAAAFCQIABQUBBQcIBwUDBgQABAAGAAIFBQECAAkDBwgGCAMEBwAHBAgAAwQEAwQJAAYECQEIAQcFBQYAAgEABQQAAAYHBQADBQkGCAAGBAYCCQYHBwEEAgYFAwkJCAgDAAIEBgYDAgMFBAkDBAcHAAMFBQYABQkGAwcFAQkICAkJAgcCAgMIBwYEBQQDBgEECQMBBgUABQcFAgMDCQYDBgAEBQAECQQGAwYHAAIBAQQECAUJAAEGBQEEBAgEBwcDCAEBCQgBAwMHBQEFBwgJBgMFCAQEAwgEBQUGBwEFAAUECAYCAQEHCQYCBwAECAICBgYGBAkGCAQFCQADCQQJCAcEAgIFAwcGBwIJAwcACQQIBgQHBgQGAgAHCAQCBgUFBgQFBQgBAQcEBgkGCAcCAgECBwcGCQgFAQEAAgAJAAEJBgUFAgAHAAIAAQMBAwQHAgQJBQIJAAMEBAYGBQICBQYBBAIFBAQIAAIEAAYCBggDCAIABAQICAQFAgIDAAEJAAkBCQYFAgIHAwgFAAgABwUFCQUABgADBAQECAcBAwQABgUBBgcDBAgIAggJBAEGBQIHCAUDCAUEAggJCAQDAwgIAgAJBAQIAQMGAQkCAQYCBwkEBgAEBAIEBQACCAMHBgQAAAUCCAIEAwACAQcHBQQHAgQHAwMBAwMGBgEHAAQHAQAGBQgEBAcJBQAHCQUBAgkFBQMJCQECBwgAAQcCBwEBBQUGBAUJBQUCAwkDCAUCAAAABQcIBQIGBAQHBAAFAQcCAwYICQgHAQIDAgkHBQYBAAcBAgYJAwYEBAMAAAYDBgEDBQYFAAkBAQgCCQUABQQGBwEBBQMICQgDCAkCBgEBAwUCBwEJCQICBwABAwkGAwcCCAEIAAcJCAUIAgYEAAYFCAMIBwUHAwcABwACAgUDBQEBBwcGAwAEAggGAwEECQkEBQICBwgABwkBAgkEBwMEAwMBCAgFAwgGAwQDAgACBwMGCAMJBwMBBwUICAIABgYEBgUHAAYGBwcDCQIFCQcBBwIEBQEDBQQFAAQEAQQECAQFCQQBAQcCAwUAAwkHBQMICAgJAQEGAwEGAwQJCQYABgkECQYGAQgJAgkECAAHAgMGAgUGAQkAAgIJCQAIAwYGBgACBgYGBQACBwgHBAYGBgUCBAIJBQIGBwABCQUFBgcHBQAFAAgEBwkIAgQHAwMHBwcIBAkFBwYFBwUBAwAECQkIAwcAAwQJAwQBBwkBBQcCBwYFBQMHBQUEAAcDAwIAAgkJAAMACAMICAQEAQgCAwUABAQDAAABAAgECQAHAAcHAQUIBwIJCQECBAEBAQgIAQMHBQIACAQICQQBAwEJCAEBAQgACQkHBgAEAgEBBwkEAgADBgYIAQcCBAICAwgHAQgFBQcHBgYCCQYGAAUFAAIBAAMFCQEHAgUHCQgJAQIIAwIDBgAIAwgGCAYACAgIAgkDAgAIBQcAAQEBCQUBAwMGAAgCAQMICAYHBQYGBwMDAgQGBQUEAgMHAQEHAwgHAgIJAwkJAAEFAQIIAQQJAwQAAAcGCQkFBQcACQMDBgkIBAICAwkEAwYHAwAECAEIBwMEBQICCQEJBAkBBgIICAcHBwcCBgAHBQgFBgACBAkCBgEBCQMBAgICAwcFBwQEBAECAwUEAgkBBQkDBgYEAAEHBwQGBwcDAAQDAgcIBwIIAAgHBAYCBwAJCQMBBQkDAQADAwUJBwkIAwUJAwcJAwQGBgUHBwYDBQMBAQkEAwkFBQcCBAkAAgEJBwEFAAUFAAMFAAMDBwQCBggFCQIBAAUFCAUDBwQEAwkGAAAJAQQDAQUDCAUBAQIJBAAFAggAAgkIAgQIAwgAAgAABAgFBAkACQUIBgkBAgcCAQYGAAYFAAQIBQUCAAMIBgMBCQIJBwMGBAgDBAgFBggDAwEEAgAECQACAQkJCAEICAMDAwgFCQYGCAcAAwAAAAUHBQAHCQEGBAgJAwUGAAQEBwgJAQgCBQUDAQEICAMABwABCAEGBAkICQgIBgkBCAQCBwcABQEBAAYEAwAEAwUBBgMDBAQFBgkIAwMBAAIBBAQECAQJAAkJBQAGAAUAAwkIAwYBCQQBBgQFBAEGCAUJBAUDCAgBAAYICQAFBgIECQMCBggFCQIDAQYDBgIHAgECAAMABAgFAAgFAgcIAQUJCAAJAAkDCQUHCQYABwEFBQkFBAYCBAUJBwQCAgkJBgIBAwcBCAcJAQgJAwcHBwEHBAQAAgQFAggJBggGAAcHBQABCQYABAkCAgYHAQAFAAcCCAYDBgEHBAgHBAEHCAIEBAQEAQgGBwIJBAUABgMICAIABwIEAAMBAAgBBwcCBgMBAgcAAAAGAgcCCAMGAwAJBwMFBAkGBAUABggEAAQEAAQBBgQDBwEDBgQBAQUGAwgDBAEHBAkICAQFBQYHAwkBAAUDBwABAQgDAgAFCQYDCAEABAYCBAgDBAIBCQEHCQUHBgEFCAgCBgQGAwEGBAADCQgEAAgEAwkHAQEHBQEJBAQCBAgGBwEDAAcDBQgBBAIDBAgICAQIAQgJAwAGAAIJBwMHAAEDBQkDAAUABQgHAwYCCAAACQMABQgBAQIGBgYIBgYDAwkFBAMDAQgBBQIGAAcJAwIHAQcGAQYCBgYHAAQJAwIFAAgGAgYHBAIBAQADBQUCBgIBCAADBAMEAgkCCAkGAgAGBAUEBwgBBgMEAgkEAwUBAgYAAQEAAgYJCQABAgQGBwYBAQcHCAcGBQEHAAAJAgEGAgIFCAUCBwgEAAkACAEDCAIIAgAGAwgHCAIGBgYABAUDBAEBCQcABgIBBAIHBwEIAwgHCQgHAgcJBwICAwYDAwkBAQQIAAkHBQIICAMGAwAIAgIFCQIGCAMCCAMDAQEDAAEIBAEFCAEFCAAHCAcFCQUFAwAHCAQFAQYEAgQEBwcJBwIFCAUIBQMACQcCAAYIBAAHAwcGCAQHAAAEAAYHCAEDCAEHAwYBAQYGCQgACAIHBwEDAAMABwgHBgUCBgcFBQgJBwMCBAgDCQUHBgIGBwMIBQEICQIGAQQCBwQABAQICQcJAgYIAgYABgMEAQUGBwUABwYCAAUBAQMHAwgIBwUJCQUFBAAHAQQGAAkJBAYIBgUHCAACBwYCBAgBAAQIAAcFCAcHBQUCBwYCAgADBgIAAwkEBQQEAQIFBwAFAAIFAAQIAAIHAgcCAwQHAQIDBAICAwkJAggABAAAAQkDCQAIAwkEBwkBBQgBAwYCBAgGAwMAAAADBgEGCAYHAwAABQgICAYBAAgHCAUIBwAHAgcBBQUIAwIIBQkJAgAGBwEICAgHAgAJBwkGCQQBBAcCBQIGAgkABwEJBwYFAwgECQMJBQcFAQMDBgYHBQMFBQgHBwQGBQUGBAgDAAkCAwICBwQABwEECAEEBwgCAAcCAAICAgIDAAUCAwIIAAAIBAYECAMFAQMJAQQJBgcEAAACBQYCBQYFAQcJAgYHAwACCAEACQkHCAQGCQcHBQgIBwUDBQUACAkHAgEEBwEAAgMCAgkEBwcABQAICAYIBQYABwMAAgIFAgAHBwcBBQIBCAUCBAgHBwYHAQAJBQMJAwUABwQCAAEGBQkDAggFBwMBAgAIAQgFBgkDAwgABAIABgQCAwQFCQEBCQQHBgQBAAAHBwAHCAAFBgQECAkHBAkFAwcCBAEJBwgBAwkDBgcIBgUGBwkJBwAGAgMEAgkFAwkCCAgGAAQBCQMDBQEFBQkCCAUFCQUEBgcDBQQCAwkECAEJAgYCBAgIBwcDCQMJBggCBgkABAQFBwAEBgQBAwEGBwgICQcABQcDBgAEAAMGCAgIAAIDAggIAQQBAQYDAgIIAAAEAQMCBwAGBwkJBgcACQkIBAgDBAAJBgYEAwgEAwYJBgQHBAcICAICCAgAAQUICQgHAQADBgYACQMJCQEFCQQBBwABAAUDCAgEAQcABQAEAQcBBwkGAwkGAgUDBQgIAgQHBwQGAAUFAgYCBQYIBQYJAQMBBAUJAAkAAwUGAQUIBQYGCQEJAQcBCAADBAACAAMDCAgAAwYIAQEDAwICAQgDAgACAwIDCQIDAAQFBwEGAQAAAwgFAQYFBAgBBwEGBAkEAgEFAwEFCAcHBwIHBQQEAgAGCAkJBQgABAQBAwcJAAAJCAUDAAcBCQcIBAAJCQECAAYCAgUBAAQIAwQIAwQAAAMGBwIEAQUCAAcJCQEFBQkBBAYBAggDBQcIAAQCBwIHBQAIBwgIBAkCBQUJCQYEAgYGAwgHBgYHBQIJAAMEBAgEAAIHAAUCBgYIAAYCAAIFBQgABwAICQACAQIFAQYCCQYECQkBBAMBAggGAwYAAgACAwMFAgMGBggFBAABAgICBgcJBgQDAwYIBQQEAAMDBQIFAQUFAQcFAgEICAAJBgAGAwIBCAMEAgECAAMACAUBBQACAAMJCAIIAQACBgIAAwcFAAYCBggEAgEIBggCCQIGCAUCBwMCAgYEBggFAgEFAggCCQAGAwYEBgcHCQYECQgABQgHBwYHBwAHBwIJBwQCAgMIAAEJCAAGAggDCAAHBgQDCAEDCAcDAAMECAgJAQMCAQQIAAQGBwMJBAkABAYFBwAIAQcFBAkFBwMCAAkIBwMHBwQHBQgFAgQGAgMICQcBAwQCCQcBAgIGAAUFBAUCCAkECQkBBAMBBQUFBggBBQUJBQkIAwEBBAYGCQMGBAQIBgUECQACCAQIBgMGAQQDAwgECQkABwkFAwEHAggECQYBBgAFCAEJAwAIAwgEAgUGCQABBAAHCQABAwkFAQQBAQQIAwkFBAQDBwcHAgMJAQkEBAAHBgYHCQAGBAkABAACBwkGAQECAAgIBwQCBwUHBgEIAQQJCAkAAQAGAAMABgQBCQcBAgIFAwQBBwQHAAIICAEECQkGAwQJAwkIBQMIBAcBBAEJBggDAgQIAQEFCAYBBgIFAAkFAwUABgUECAMIAAIJCQEFAwIGBQMAAwAHBAACAQkFAQcFBQUJBAgEBgYEBgQBAQICCAQJAAUIAQAIAwIECAMACQMECQcCBgICCAkGAQQHBwkHAwIJAgcDAAIGBwIDAwEJBQgCAwUAAAQFCAMFAAcHBgAJBQEEBAUDAAYBBAYFBAgFBwMACAMFAgQJCAUJCAEJBQABBQMDAwMEBAMGBwgACQYDAAQCBwMFBwMHCAQGAgIDAQAHBQYGBAEBBAcCAwUFAAQEBAgJCQIGCQgDBQMBBAUAAQgFCAYABwgJAgIGAwICAggJAQkHBAUGAwIEBQYGBQgGAgIJBAkIAwACBwcFCQcHAwgHBQkJCAIBAAgAAgYHCQICBQQDAgEBCQYFBQQJCAkGBQQGAAUHCAQICAcECQQJBgkHAwkJAwYAAggHCQQAAAQBAAMIBAgEBAMDBQcHCAUGAwIBCAMIBQQCAwAIAwMCCQAJCQABCAMAAwgGCAUBAwABCQMFAAYCAwEFAwIIBQcIAAEBAAYEBQkABwIDBgQGAwYJBAkACAMFBgIAAQMJBAIFAAUDAgUAAAAAAAkBAwQDBQUGBgEECQcEBAcFBwEGCAcJBQcAAgkCBQAHBgQFAgkEAwIFCQIFCQcDCAMFBgUDAAcGCQUBAwYJAwYCBQMECQkBAgICBAYIBwkJBAgICQAHAAQDCAYEAgABBAACAQUEBgYAAwQFBggHBgADAQgICAkJBQUIBgUEBAIFBgkCCQEABAYGCQgABAkBCQQHAwUEAwQGAQYCBAMHCAQFBQkABgMEBQUFBwMBAwMHAAcHBggDBQUIAQkEAAQCAQYAAQYFAAgBAQcFAgUJAAgDBAQCBwAHBgYCAwcBBgQDAwAABgQIBwADBgQHAwMGAAIEBAEIAQkBAgYEAgUDAwUDBQgDCQcDAAABAAEABwcECAkAAQUECQQHCQkHAAEACQkFAQcCCQEFBggBBwMACAQHAAgJCAgEBQYEBQkAAAUDBwcEBAcEBQgDBwIHAwgICAcIBgAFBwkIBQYABgQEAggJCQEJBQgEBAgECAcECAMABwgIAwkJBQkCBgkHBQcECQkCAwkFAwUGAQIJAQQIBQEHCQADAwYJAAQGAAACCQcBCQgFBggFCQUCBAIFAgIFBAIHAQIGCAIFBgIGCQUFAQUCBAQCBAEJAwcCBwQCCQcEBgYBBwEEBwYBBwMAAwYECQAEAQcJAAIHBgYHAQkFAgkIBgQAAQMCAggBAAgBCAEAAQQHAwcHAwMJCQAIAgUICQUDBQQBBgkCBgIGAQgBAwMAAgEHAAUEBQMGAgEGAAcDBgYDBgEFAgkACAAEBAABBwUDAwMCAAMFAgYCBQEEBgIBCQMEBgkFCAAGAwYACAECAwMBBQUDAgYHAQMCAwYBAgIIAAgEBwADAQMIAwQBAgADBgcHCAMGCQUGBgkFBAMIBQMIAwcJBgIFAQAFAgMFBwkGAAcBAgAIAgICBwgAAgQECQcDAgUEBQEJBQAFAgQBBwkAAwMECQEAAQkFBggCAgEEAAYHCQABAQQDBgkIAQMHAwkJBQQCAwYICAUFAQgIBQcJCQMEBAIHCQYGBwUEBwYBBAICBgkAAAUBAQcIAwYGAgACBAkFAgQGBAYBBQIHCAYHBwAIBAMGCQECAQcEAwQJCQICBAMABQcFAgcICQkCBwgHBwUIBwYGCQABCAcFBAUDBwYGBQUGCQkBCQUJBAQBBQgDBQEIBgUAAgMHAQEBCAcFBgMEBwAABQMECAYFAgcDAgMCBwgACQAJBQgABwMJAAkHAwAGBQcIAgMBCQACCQEDBAADAAYCAwgGAgkFBgEIAgQBBgIJBAYHCQYDBQQFCQgFAwABAgYEAQkEBAYFBQQDAgABBAkJBwYBAgYGAgYCBQkECAIGAQUJBgEICQMBBwEBAQkBAwgJCQEJBgAFCAUFBQMICQAHCAIJAgABBQAHCAAACQEBCAIIBwIABwIAAAUBAAgGAwcCAQkHBAEDCQIEAAMHBQEBCQUBBQQGBgEFAgEGBgUHCQQDBAEHCQAFBwMDCAQBCAIFBAQICQIJAwIGAQEGBgEECQgFCQEFBgkDAgEHBAkICAkFBQMIAgADBgEBCQAJBwUDAgkFBQQCCQkDAAYDAQcECQIIBQgGAAEDAgIDBgYAAggDAggGAQgEBQQCAwkABAQCBQQCBwYHAwMGCQYIBQYGAwYHAgUDAQEJBwUEAgEGBAUJAgUFAwMACQICAgMCBAkCAwAEAgAEAQMHAgQCAgMCAwAIAwUBAggIAgECAQEJBgUJBggAAwYDAQMHBQYBAQIABQUAAQYFAggJBwkBAQMDBQYHBQkECQMGBgAIBQYFCQgDCAEDAQMFBgUEBAkGBwEECAMHBwYHCQQHBwABBAEECAIEBwYHAQgHBQEDAQADBAUECAYDBAkDAQUDAAcAAwQFAQYGAggICQgGBwIEAgUBCQYDCAcBCAECAAkCBAQEBgEDBQIJBwUCBwUFBAQIAAcACAACAQkHBwMIAAMACQgFBgQIAQUFBwABBgEDAgYEBgEHBgICBgEIAgIGAwAICAQGAgAHAAgJAQYGBwAIAQgCBAYHBQcJCQMJBAQGBAEECAQCCAYCAgkBCAAGAgYAAwYGAQMGAwMBBwgBAAcGBQAIAgYFCAcIBAUJBQUBAgQHAgkHCAMHBAYDCAgEBQEIAgYCAgkICQACBQkDBQkIBgMJBQkDAAgGAAYDBQgABgEIBgkAAAcGBggHBggIAgADBgcBAwcEAAcBBgUCAAcJAAMIAgEBCAgDBQcFBgcEAwkDCAkABAAEBQcABAMJBQcCAgUIAwEDBQQJBAMJCQIACAAJAQkJAAcBBAUAAwgHBAUFCQMCAAAJAAUAAQMHCQEFCQYACQcCCQIBAgcFAQAHAwQHBgQGCAUHCQUHAwYDCQMEAQUGCQcEAAYFCAAHBwMCAAQIAgAGCQgJAgUABAACBgMBCAcEAQkBAggFBgMAAAIGAAcCAQgECAcJBAEGBQcHAwIFBwMGBwMCAgkGBwMJAQIBCQcCAwMIAQgDBQAFAwgFBwQJAAYIAwIDAwQDBwgHAQMHCAkFBQICBAUACAADBgAICAkGCQEFAQIBAwICAgQFBwUJAgAFBwAFBgUGBgQABwEFAgkIAQcGBAQABwEIBQcIBQUBCAUJBwcGBQICAQIEBAEGAwkBCAIDBwYJBQcEAwkFBgICBwMJAwAFCQAJBgYGBQkBBgUCAAIDAQEBBwYCBgUJCAcJAwIHCQIAAQEBBAgACAAJBQYHAAgEBAQCAgIFBwIJCQICAgAHBAUCAwMBCQgBCAYGAwkBCQcCCQAJAwAEAgMEBgYJBgEBAAUGBwAEAQEJAAABCQYHAwUEAQIABAADAAIDBQkFCAkDAQcICQQBAQkABwAACQYAAgkGCAUFCQMDBQcBAgMDBQYGBAEBCQEGBgYCAgYGBgMFAAMABQEBCAIFAwUBBAQDAQIHCQYDCAUGBgkBBQcHBAICBwUEAQUHBwgCCAMHBAYCCQkAAAcDAQAIAgQEBgcBCAYAAwQAAwkIAAAIAwkCAQUEBQAIBAIFCQMAAgcIBQkFAAAGCQcCCQICBgQAAQQDCQQAAwQEBgkEBwIFBAMEBQcAAgEBCAcDAgYFAgMICAgJAwYJBgADBgQCCAgFCQIHAQkAAwMDAgMJCQUDAgQCBAYEBQMECQgFCAMEAQQICQIGAgMFAAMFAgUDCQECCQgJAwgIBQUIBAQACAMGBAkGAQAFAQQACQIGAwYHBAYEAwgFAQgDBAIFBgAHCAcEBAICCQMFAAkJAAEDBwEJBgUCBgkFBwkEAggDAQcICAYIAAUFBwEJBgIBBQEJBgEGAwIFAwkDBQQABgkFBAIDCAUIAwYEAQQJBQgJBwMGCQUJAQEBCAQBBAYJCQgJBwQEAQkJAQkDAgMHAQEDBggDBAMDCAMHAQADCAQHCQgAAwAEBQcGBwMFAAIDAwMABgQEBAQECQkABwQAAgIBCQUJCQAFCAYHAwkFAggJBQgAAAMABgMHCAABAgIDAQEDAgIFBAMJAwMCAwcCCQIBBgcFBQkCBgkHCQcEAQgDAwgACAcABwcJBgMAAwcEAQUCAAUIBgcAAQAFBgIBBAEIBwkJBwMCAgQHBgMJBgMJBQEGCQIHBAMFBwIHAggHBQQFCQQDBQACBQUDAQkDBwEAAwQEAgUHAgEBBwcHAAEGCAUCBQMHCAcHAAgFBgYFBQYHBwkHAQAACAAEAAYEBQUABQMGBAUABwEIBwEDAgAGBwMHBAgAAwkEAAMACAcDAwYFAAEFAAkJAQMABgEGBgUHAgQACQIGCAIIAwcHAwMJBQcGAAkEBQAHAgkFBQgHAQkFBgMGAwACBwYEAgUEAQUHCQcDCQgABQEJBgEABggABwYJCAgEBAQBBwcBAQYHBgMCCAcIBwkBAQMICQUJAQAGAwQEBAcBAwABBQQECQIEAgIGBwcGBwEEAAgDAwUGAAgDAAIAAQgEAAIJBQQEBQIBBgQGBQgABwMCBgcCAgMGBgIAAwIFAQcFAgUHBwkJCQIACQIABQUFAAUABgYABgkJAgQFBAABAAgJBQcDAAIEAQMIAQgCAAYJCAUBAgQACQYICAAHAQYJCAkJBAkGAwMEBQAIAAQAAwAAAAMDBQkEBwEFBAEGAQcJBAMEBQIJBAkICAcEAgMJAwYHAAIHAAYBAAUHAAUJBAMCCAUIAQkGBwYHAwYEBQEICAgHBAYFAQUJBgAJBQgFBwMJCAMEBgUHBQEBBQQBCAUIAgkGAwMEAAIEAQEGBwMJAAIICAgGBwcBAgYGAQYAAgUIAgQCAwEFBQQEAAUEBwYDBAUFBwUHAAEBCQcFAAEJBgMGBQQDBgUHAgICAAUGBgMHCAYCBgEBBwUCBQMECAUICAYBAgMFBgkABAUCAwQJBQUJBAYEAQEICQQEAgMDAAcBCAgFAAAHBwEAAQQBCQAEAQEABwMHAwkAAQgJAAAGAAkBBgECBwgABQYIAgUFAQUJAQgFAwEGBgAHAwUECAAHAQIDCQgDCAcHCQcACQQACQYBCQgGAwcECQUEBwECAQIDCQIABQAABQUEAAkJAQYHBwcEBQECAAYDCAkIBAEACAMJCAgHCAMAAAcECAYJCQQFBwICAQUDBAAFBQAIBAcICQQCAQcEBAQJBAMGAAcHCAAJAQIGAgkHAQkHAwMBAgMFBAMHBAEEBgYHBQQGBwkIAQgBAgIFBQYBBAYGAQIABwYICAIIBAUAAAEHCQcECQQAAggHAQAIBwYIBAgCCAABBQMIAQcECQMHCQUBCAMACQMGAgQCBwYCCQgIAQUHBwEJBQQBAAcIBwMHBgIFBQQGBwIGAwIABAcIBgYAAQUDCAMJBwEHAQADCAUAAAYEAQECCQAGAgEDAgcGAQgFBwIABQAEBgICBQQBCAIBBQMICAkJAAEIAQQJAAMBAQQHCAMECQYGCAMIAgkEBwcAAwACBQcDBwQGAAkDBgUEAQIBCAEJBAgJCQMDBwAGBgEGBgYEAgIHAAYBAgAAAQYAAwUHCQEGAwcFAQIHBwUIAAkBAAEGCQEGBQMDAAcBAgcJAwcEBAQGBwQHAwEHCAgHAAEEAAkHCQkBAgAEBwQDBAEHBQcACQAAAwEEBwkDBgcIBAcEAAgDAAEHCQYJAQQGBwEABgMHAwEIBgUJBQgBAAkCBgIBAQQHAwAHBQgJBQQABAcCAwgDCAEBAgQIAQYIAQgFAwIHAgIECQYABAkJBgkJBgcDBAMDCQQHBwgBAQYGBQcJBAcGBwAAAgkJBwcFAAEEAwQGBAAJCAcIBAQICQcBBgcJCQABBQEBAwYHCQgDAwgFAgQDBggGCQAGCQABBggBCAIBCAEHAQkBCAAJBAEACAQEBggCAgUFAAYGCQgAAQAABAgDBAcHBQkCBQQECQkECAcGCQMEAAQIAAEJAgEGBwEIBQAEBgQBAQkICAkBAgYABQIFBggDBgEIBgEFAAMDAQkCCAcIAgMGAgUACQAHAQEHCQYBBgcFBAYDAwYBBgAJCQgCCQYCBQQHAgYFBgAJAwcAAgMFBgQDBAYJCAgDAgAAAwkCAggBCQIECQMGAgAGBwAFAAYDAQMACQUEAwcABgIJAAcGBQYIAgEABgYFCQMFAAYABwQGCAMDCAYDCAkEAAgIAwgABwcAAwYIAQUJBgIICQABAAUFAgIIAgMGBwQFAwcEBQgIAwkFAAYCBgMGBgkCBQYFBgMBBwgABAcDAwgGAAkCCAUECQYGBAIAAQEJAwgCAQkFCAQIAgMFAwYECQIGAgkIAAUJAgQFAgEABwQEAgcEAAYGAQUEAAEJCQUGBQIJBgYEBQQICQAJBQUFBwkDAwgEAwkDAQACBQIIBwkHBwcJAgAABgcGAQgAAwgFAAUGCQEBAgYICQcHBQgGAQQAAQgEAwQHCQMHBQUACQECAQIHBgAJAwIBAQcIAQYFCQAACQUDCAgCCAUHAAcACQYJAgkECAUCAAEGBgEHBgEHBwgFBwAGBAIBBggAAAIHBQkBAAQDBwYGAwEIBAcCBQkDAQgBCQIJBgICAQgDCQgCAAQIBAUBAgIGBAEIBQcHBwMFBQcJCQgCCAIBAwMECAkEAAcGAggBAAYBBgUCCQcCCAcDBwgBAgEHBAgIBQACBgYBAgYFAAEFBQIFAQkIBwECBwAJAwEJAAQGBgUJCQQCBwEIAQQCAwcIBAYEBQECCQMFCAABCQYHAwcEBAAGAgEBAwAHAAUJCAkDBwkJBAgIAAEGBAAGCAEGAQYDAgcIBwAEAwQEBQcJAgUCBAUAAQcGBwYAAAAFAQECBAYJCAEGAQYIBQQABAAECQMIBQIIBgIIAAQEAAgHBQIICQIJBQkCAQAEAQcHAQcDBQgHAgUBBQIABgABAgQECAADAgADBQEDAwAFAQgDAQgHCAQCAAAHCAEDCQACCAEECAgDAAgFBwUEBwcGAQgBCQcAAwEGBQUDCAkEAAkFAQcCAQUCCAMDBgcBAwQEBQcABgcDAwcICAIFBgkHAAIJAwAFBQYJAQQBBQYEAQQAAwMGAwAIBgcFBQcABwkEBAIJAQcBCAgIBQgDCAkDAAACAQgEBAAACQUDAQUBAwUAAgYBBQMACAYJBgkGAAQBBgYECQEECQcIAgYCCQcIBQYJBgAACAEDAgcFCAQFBggJAQkBCAEF
```

</details>

### Environment

Zed 0.161.1 – /Applications/Zed.app
stable
8 GB
aarch64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary>

<!-- Click below this line and paste or drag-and-drop your log-->
```
2024-11-14T02:15:54.175875-03:00 [INFO] ========== starting zed ==========
2024-11-14T02:15:54.190949-03:00 [INFO] Opening main db
2024-11-14T02:15:54.200934-03:00 [INFO] Opening main db
2024-11-14T02:15:54.23625-03:00 [INFO] Using git binary path: Some("/Applications/Zed.app/Contents/MacOS/git")
2024-11-14T02:15:54.316693-03:00 [INFO] set environment variables from shell:/bin/zsh, <redacted>
2024-11-14T02:15:54.697825-03:00 [INFO] no tokio runtime found, creating one for Reqwest...
2024-11-14T02:15:54.769103-03:00 [ERROR] theme not found: Catppuccin Mocha - No Italics
2024-11-14T02:15:54.775581-03:00 [INFO] initializing context server client
2024-11-14T02:15:54.77713-03:00 [INFO] Registering built-in prompt template: content_prompt
2024-11-14T02:15:54.779498-03:00 [INFO] Registering built-in prompt template: project_slash_command
2024-11-14T02:15:54.779627-03:00 [INFO] Registering built-in prompt template: suggest_edits
2024-11-14T02:15:54.779875-03:00 [INFO] Registering built-in prompt template: terminal_assistant_prompt
2024-11-14T02:15:54.780395-03:00 [INFO] No prompt template overrides directory found at <redacted>. Using built-in prompts.
2024-11-14T02:15:54.78677-03:00 [INFO] extensions updated. loading 3, reloading 0, unloading 0
2024-11-14T02:15:54.886169-03:00 [INFO] Opening main db
2024-11-14T02:15:54.910371-03:00 [ERROR] theme not found: Catppuccin Mocha - No Italics
2024-11-14T02:15:54.913765-03:00 [ERROR] theme not found: Catppuccin Mocha - No Italics
2024-11-14T02:15:54.922174-03:00 [ERROR] theme not found: Catppuccin Mocha - No Italics
2024-11-14T02:15:55.046501-03:00 [INFO] building git repository, `.git` path in the worktree: ".git"
2024-11-14T02:15:55.061051-03:00 [INFO] Opening main db
2024-11-14T02:15:55.136125-03:00 [INFO] Opening main db
2024-11-14T02:15:55.178692-03:00 [ERROR] No path or contents found for buffer
2024-11-14T02:15:55.178821-03:00 [ERROR] No path or contents found for buffer
2024-11-14T02:15:55.264205-03:00 [INFO] using project environment variables from CLI. <redacted>
2024-11-14T02:15:55.591384-03:00 [ERROR] {
  "thread": "main",
  "payload": "not enough space in Arena",
  "location_data": {
    "file": "/Users/administrator/actions-runner-2/_work/zed/zed/crates/gpui/src/arena.rs",
    "line": 80
  },
  "backtrace": [
    "backtrace::capture::Backtrace::create",
    "backtrace::capture::Backtrace::new",
    "zed::reliability::init_panic_hook::{{closure}}",
    "std::panicking::rust_panic_with_hook",
    "std::panicking::begin_panic_handler::{{closure}}",
    "std::sys::backtrace::__rust_end_short_backtrace",
    "_rust_begin_unwind",
    "core::panicking::panic_fmt",
    "gpui::element::AnyElement::new",
    "editor::display_map::HighlightedChunk::highlight_invisibles::{{closure}}::{{closure}}",
    "editor::element::LineWithInvisibles::from_chunks",
    "editor::element::layout_line",
    "<editor::element::EditorElement as gpui::element::Element>::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "<editor::element::EditorElement as gpui::element::Element>::prepaint::{{closure}}",
    "<editor::element::EditorElement as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint::{{closure}}",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::window::WindowContext::draw",
    "gpui::window::Window::new::{{closure}}::{{closure}}",
    "gpui::window::Window::new::{{closure}}",
    "gpui::platform::mac::window::step",
    "<gpui::platform::mac::platform::MacPlatform as gpui::platform::Platform>::run",
    "gpui::app::App::run",
    "zed::main",
    "std::sys::backtrace::__rust_begin_short_backtrace",
    "std::rt::lang_start::{{closure}}",
    "std::rt::lang_start_internal",
    "_main"
  ],
  "app_version": "0.161.1",
  "release_channel": "Zed",
  "os_name": "macOS",
  "os_version": "14.2.0",
  "architecture": "aarch64",
  "panicked_on": 1731561355591,
  "system_id": "6eed55d8-a267-4d49-b7f4-4c3096e8f6aa",
  "installation_id": "de095602-7b02-4dae-a2f5-90f37f587689",
  "session_id": "a96ee99e-9e13-40dc-8696-568688639465"
}

```
<!-- Click above this line and paste or drag-and-drop your log--></details>

--- a/crates/editor/src/display_map.rs
+++ b/crates/editor/src/display_map.rs
@@ -66,7 +66,7 @@ use std::{
 use sum_tree::{Bias, TreeMap};
 use tab_map::{TabMap, TabSnapshot};
 use text::LineIndent;
-use ui::{div, px, IntoElement, ParentElement, SharedString, Styled, WindowContext};
+use ui::{px, SharedString, WindowContext};
 use unicode_segmentation::UnicodeSegmentation;
 use wrap_map::{WrapMap, WrapSnapshot};
 
@@ -541,11 +541,17 @@ pub struct HighlightStyles {
     pub suggestion: Option<HighlightStyle>,
 }
 
+#[derive(Clone)]
+pub enum ChunkReplacement {
+    Renderer(ChunkRenderer),
+    Str(SharedString),
+}
+
 pub struct HighlightedChunk<'a> {
     pub text: &'a str,
     pub style: Option<HighlightStyle>,
     pub is_tab: bool,
-    pub renderer: Option<ChunkRenderer>,
+    pub replacement: Option<ChunkReplacement>,
 }
 
 impl<'a> HighlightedChunk<'a> {
@@ -557,7 +563,7 @@ impl<'a> HighlightedChunk<'a> {
         let mut text = self.text;
         let style = self.style;
         let is_tab = self.is_tab;
-        let renderer = self.renderer;
+        let renderer = self.replacement;
         iter::from_fn(move || {
             let mut prefix_len = 0;
             while let Some(&ch) = chars.peek() {
@@ -573,30 +579,33 @@ impl<'a> HighlightedChunk<'a> {
                         text: prefix,
                         style,
                         is_tab,
-                        renderer: renderer.clone(),
+                        replacement: renderer.clone(),
                     });
                 }
                 chars.next();
                 let (prefix, suffix) = text.split_at(ch.len_utf8());
                 text = suffix;
                 if let Some(replacement) = replacement(ch) {
-                    let background = editor_style.status.hint_background;
-                    let underline = editor_style.status.hint;
+                    let invisible_highlight = HighlightStyle {
+                        background_color: Some(editor_style.status.hint_background),
+                        underline: Some(UnderlineStyle {
+                            color: Some(editor_style.status.hint),
+                            thickness: px(1.),
+                            wavy: false,
+                        }),
+                        ..Default::default()
+                    };
+                    let invisible_style = if let Some(mut style) = style {
+                        style.highlight(invisible_highlight);
+                        style
+                    } else {
+                        invisible_highlight
+                    };
                     return Some(HighlightedChunk {
                         text: prefix,
-                        style: None,
+                        style: Some(invisible_style),
                         is_tab: false,
-                        renderer: Some(ChunkRenderer {
-                            render: Arc::new(move |_| {
-                                div()
-                                    .child(replacement)
-                                    .bg(background)
-                                    .text_decoration_1()
-                                    .text_decoration_color(underline)
-                                    .into_any_element()
-                            }),
-                            constrain_width: false,
-                        }),
+                        replacement: Some(ChunkReplacement::Str(replacement.into())),
                     });
                 } else {
                     let invisible_highlight = HighlightStyle {
@@ -619,7 +628,7 @@ impl<'a> HighlightedChunk<'a> {
                         text: prefix,
                         style: Some(invisible_style),
                         is_tab: false,
-                        renderer: renderer.clone(),
+                        replacement: renderer.clone(),
                     });
                 }
             }
@@ -631,7 +640,7 @@ impl<'a> HighlightedChunk<'a> {
                     text: remainder,
                     style,
                     is_tab,
-                    renderer: renderer.clone(),
+                    replacement: renderer.clone(),
                 })
             } else {
                 None
@@ -895,7 +904,7 @@ impl DisplaySnapshot {
                 text: chunk.text,
                 style: highlight_style,
                 is_tab: chunk.is_tab,
-                renderer: chunk.renderer,
+                replacement: chunk.renderer.map(ChunkReplacement::Renderer),
             }
             .highlight_invisibles(editor_style)
         })

--- a/crates/editor/src/element.rs
+++ b/crates/editor/src/element.rs
@@ -16,8 +16,8 @@ use crate::{
     items::BufferSearchHighlights,
     mouse_context_menu::{self, MenuPosition, MouseContextMenu},
     scroll::scroll_amount::ScrollAmount,
-    BlockId, CodeActionsMenu, CursorShape, CustomBlockId, DisplayPoint, DisplayRow,
-    DocumentHighlightRead, DocumentHighlightWrite, Editor, EditorMode, EditorSettings,
+    BlockId, ChunkReplacement, CodeActionsMenu, CursorShape, CustomBlockId, DisplayPoint,
+    DisplayRow, DocumentHighlightRead, DocumentHighlightWrite, Editor, EditorMode, EditorSettings,
     EditorSnapshot, EditorStyle, ExpandExcerpts, FocusedBlock, GutterDimensions, HalfPageDown,
     HalfPageUp, HandleInput, HoveredCursor, HoveredHunk, JumpData, LineDown, LineUp, OpenExcerpts,
     PageDown, PageUp, Point, RowExt, RowRangeExt, SelectPhase, Selection, SoftWrap, ToPoint,
@@ -34,8 +34,8 @@ use gpui::{
     FontId, GlobalElementId, Hitbox, Hsla, InteractiveElement, IntoElement, Length,
     ModifiersChangedEvent, MouseButton, MouseDownEvent, MouseMoveEvent, MouseUpEvent, PaintQuad,
     ParentElement, Pixels, ScrollDelta, ScrollWheelEvent, ShapedLine, SharedString, Size,
-    StatefulInteractiveElement, Style, Styled, TextRun, TextStyle, TextStyleRefinement, View,
-    ViewContext, WeakView, WindowContext,
+    StatefulInteractiveElement, Style, Styled, TextRun, TextStyleRefinement, View, ViewContext,
+    WeakView, WindowContext,
 };
 use gpui::{ClickEvent, Subscription};
 use itertools::Itertools;
@@ -2019,7 +2019,7 @@ impl EditorElement {
             let chunks = snapshot.highlighted_chunks(rows.clone(), true, style);
             LineWithInvisibles::from_chunks(
                 chunks,
-                &style.text,
+                &style,
                 MAX_LINE_LEN,
                 rows.len(),
                 snapshot.mode,
@@ -4372,7 +4372,7 @@ impl LineWithInvisibles {
     #[allow(clippy::too_many_arguments)]
     fn from_chunks<'a>(
         chunks: impl Iterator<Item = HighlightedChunk<'a>>,
-        text_style: &TextStyle,
+        editor_style: &EditorStyle,
         max_line_len: usize,
         max_line_count: usize,
         editor_mode: EditorMode,
@@ -4380,6 +4380,7 @@ impl LineWithInvisibles {
         is_row_soft_wrapped: impl Copy + Fn(usize) -> bool,
         cx: &mut WindowContext,
     ) -> Vec<Self> {
+        let text_style = &editor_style.text;
         let mut layouts = Vec::with_capacity(max_line_count);
         let mut fragments: SmallVec<[LineFragment; 1]> = SmallVec::new();
         let mut line = String::new();
@@ -4398,9 +4399,9 @@ impl LineWithInvisibles {
             text: "\n",
             style: None,
             is_tab: false,
-            renderer: None,
+            replacement: None,
         }]) {
-            if let Some(renderer) = highlighted_chunk.renderer {
+            if let Some(replacement) = highlighted_chunk.replacement {
                 if !line.is_empty() {
                     let shaped_line = cx
                         .text_system()
@@ -4413,42 +4414,71 @@ impl LineWithInvisibles {
                     styles.clear();
                 }
 
-                let available_width = if renderer.constrain_width {
-                    let chunk = if highlighted_chunk.text == ellipsis.as_ref() {
-                        ellipsis.clone()
-                    } else {
-                        SharedString::from(Arc::from(highlighted_chunk.text))
-                    };
-                    let shaped_line = cx
-                        .text_system()
-                        .shape_line(
-                            chunk,
-                            font_size,
-                            &[text_style.to_run(highlighted_chunk.text.len())],
-                        )
-                        .unwrap();
-                    AvailableSpace::Definite(shaped_line.width)
-                } else {
-                    AvailableSpace::MinContent
-                };
+                match replacement {
+                    ChunkReplacement::Renderer(renderer) => {
+                        let available_width = if renderer.constrain_width {
+                            let chunk = if highlighted_chunk.text == ellipsis.as_ref() {
+                                ellipsis.clone()
+                            } else {
+                                SharedString::from(Arc::from(highlighted_chunk.text))
+                            };
+                            let shaped_line = cx
+                                .text_system()
+                                .shape_line(
+                                    chunk,
+                                    font_size,
+                                    &[text_style.to_run(highlighted_chunk.text.len())],
+                                )
+                                .unwrap();
+                            AvailableSpace::Definite(shaped_line.width)
+                        } else {
+                            AvailableSpace::MinContent
+                        };
 
-                let mut element = (renderer.render)(&mut ChunkRendererContext {
-                    context: cx,
-                    max_width: text_width,
-                });
-                let line_height = text_style.line_height_in_pixels(cx.rem_size());
-                let size = element.layout_as_root(
-                    size(available_width, AvailableSpace::Definite(line_height)),
-                    cx,
-                );
+                        let mut element = (renderer.render)(&mut ChunkRendererContext {
+                            context: cx,
+                            max_width: text_width,
+                        });
+                        let line_height = text_style.line_height_in_pixels(cx.rem_size());
+                        let size = element.layout_as_root(
+                            size(available_width, AvailableSpace::Definite(line_height)),
+                            cx,
+                        );
 
-                width += size.width;
-                len += highlighted_chunk.text.len();
-                fragments.push(LineFragment::Element {
-                    element: Some(element),
-                    size,
-                    len: highlighted_chunk.text.len(),
-                });
+                        width += size.width;
+                        len += highlighted_chunk.text.len();
+                        fragments.push(LineFragment::Element {
+                            element: Some(element),
+                            size,
+                            len: highlighted_chunk.text.len(),
+                        });
+                    }
+                    ChunkReplacement::Str(x) => {
+                        let text_style = if let Some(style) = highlighted_chunk.style {
+                            Cow::Owned(text_style.clone().highlight(style))
+                        } else {
+                            Cow::Borrowed(text_style)
+                        };
+
+                        let run = TextRun {
+                            len: x.len(),
+                            font: text_style.font(),
+                            color: text_style.color,
+                            background_color: text_style.background_color,
+                            underline: text_style.underline,
+                            strikethrough: text_style.strikethrough,
+                        };
+                        let line_layout = cx
+                            .text_system()
+                            .shape_line(x, font_size, &[run])
+                            .unwrap()
+                            .with_len(highlighted_chunk.text.len());
+
+                        width += line_layout.width;
+                        len += highlighted_chunk.text.len();
+                        fragments.push(LineFragment::Text(line_layout))
+                    }
+                }
             } else {
                 for (ix, mut line_chunk) in highlighted_chunk.text.split('\n').enumerate() {
                     if ix > 0 {
@@ -5992,7 +6022,7 @@ fn layout_line(
     let chunks = snapshot.highlighted_chunks(row..row + DisplayRow(1), true, style);
     LineWithInvisibles::from_chunks(
         chunks,
-        &style.text,
+        &style,
         MAX_LINE_LEN,
         1,
         snapshot.mode,

--- a/crates/gpui/src/text_system/line.rs
+++ b/crates/gpui/src/text_system/line.rs
@@ -44,6 +44,21 @@ impl ShapedLine {
         self.layout.len
     }
 
+    /// Override the len, useful if you're rendering text a
+    /// as text b (e.g. rendering invisibles).
+    pub fn with_len(mut self, len: usize) -> Self {
+        let layout = self.layout.as_ref();
+        self.layout = Arc::new(LineLayout {
+            font_size: layout.font_size,
+            width: layout.width,
+            ascent: layout.ascent,
+            descent: layout.descent,
+            runs: layout.runs.clone(),
+            len,
+        });
+        self
+    }
+
     /// Paint the line of text to the window.
     pub fn paint(
         &self,

--- a/crates/gpui/src/text_system/line_layout.rs
+++ b/crates/gpui/src/text_system/line_layout.rs
@@ -29,7 +29,7 @@ pub struct LineLayout {
 }
 
 /// A run of text that has been shaped .
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 pub struct ShapedRun {
     /// The font id for this run
     pub font_id: FontId,

